third_party/java/arcs/build_defs/internal/plan.bzl

```diff
@@ -1,37 +1,134 @@
 """Arcs Plan Generation Rules"""
 
-def _recipe2plan_impl(ctx):
-    args = ctx.actions.args()
+load(":kotlin.bzl", "arcs_kt_library")
+load(":manifest.bzl", "arcs_proto_plan")
+load(
+    ":tools.oss.bzl",
+    "arcs_tool_recipe2plan",
+    "arcs_tool_recipe2plan_2",
+)
+load(
+    ":util.bzl",
+    "manifest_only",
+    "replace_arcs_suffix",
+)
+
+def arcs_kt_plan(
+        name,
+        arcs_sdk_deps,
+        srcs = [],
+        data = [],
+        deps = [],
+        platforms = ["jvm"],
+        visibility = None):
+    """Converts recipes from manifests into Kotlin plans.
+
+    Example:
+
+      Direct dependency on this target is required for use. This rule depends on the output from arcs_kt_schema.
+
+      ```
+          arcs_kt_schema(
+            name = "foo_schemas",
+            srcs = ["foo.arcs"],
+          )
+
+          arcs_kt_plan(
+            name = "foo_plans",
+            srcs = ["foo.arcs"],
+            deps = [":foo_schemas"],
+          )
+
+          arcs_kt_library(
+            name = "arcs_lib",
+            srcs = glob("*.kt"),
+            deps = [":foo_plans"],
+          )
+      ```
+
+    Args:
+      name: the name of the target to create
+      arcs_sdk_deps: build targets for the Arcs SDK to be included
+      srcs: list of Arcs manifest files
+      data: list of Arcs manifests needed at runtime
+      deps: list of dependencies (jars)
+      platforms: list of target platforms (currently, `jvm` and `wasm` supported).
+      visibility: visibility of the generated arcs_kt_library
+
+    Returns:
+      Dictionary of:
+        "outs": output files. other rules can use this to bundle outputs.
+        "deps": deps of those outputs.
+    """
+    outs = []
 
-    output = ctx.actions.declare_file(ctx.label.name + ".jvm.kt")
+    for src in srcs:
+        genrule_name = replace_arcs_suffix(src, "_GeneratedPlan")
+        out = replace_arcs_suffix(src, "_GeneratedPlan.kt")
+        outs.append(out)
+        rest = [s for s in srcs if s != src]
+        arcs_tool_recipe2plan(
+            name = genrule_name,
+            srcs = [src],
+            outs = [out],
+            deps = deps + data + rest,
+        )
 
-    args.add_all("--package-name", [ctx.attr.package])
-    args.add_all([ctx.file.src, output.path])
+    deps = manifest_only(deps, inverse = True)
 
-    ctx.actions.run(
-        inputs = [ctx.file.src],
-        outputs = [output],
-        arguments = [args],
-        executable = ctx.executable.compiler,
+    arcs_kt_library(
+        name = name,
+        srcs = outs,
+        platforms = platforms,
+        visibility = visibility,
+        deps = arcs_sdk_deps + deps,
     )
+    return {"outs": outs, "deps": arcs_sdk_deps}
 
-    return [DefaultInfo(files = depset([output]))]
-
-recipe2plan = rule(
```

Comment:
Removing this from here will cause problems importing the PR. Do you want to leave this here for now in addition to `rules.bzl`? That way you can import the PR first, make change to tools.google3.bzl internally and then subsequently delete it?

Reply:
I like that plan of attack.